File: Algorithms/Python/652 Find Duplicate Subtrees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findDuplicateSubtrees(self, root: TreeNode) -> List[TreeNode]:
        def serialize(node):
            if not node:
                return '#'
            
            key = str(node.val) + ',' + serialize(node.left) + ',' + serialize(node.right)
            cache[key] = cache.get(key, 0) + 1
            if cache[key] == 2:
                ans.append(node)

            return key
        
        ans = []
        cache = {}
        serialize(root)
        return ans

File: Algorithms/CPP/652 Find Duplicate Subtrees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> findDuplicateSubtrees(TreeNode* root) {
        serialize(root);
        return ans;
    }

    string serialize(TreeNode* node) {
        if (node == NULL) {
            return "#";
        }

        string key = to_string(node->val) + "," + serialize(node->left) + "," + serialize(node->right);

        if (++cache[key] == 2) {
            ans.push_back(node);
        }

        return key;
    }

private:
    vector<TreeNode*> ans;
    unordered_map<string, int> cache;
};
